### PR TITLE
[torch.compile] Fix RMSNorm + quant fusion in the non-cutlass-fp8 case, rename RedundantReshapesPass to NoopEliminationPass

### PR DESCRIPTION
**Pull request by ProExpertProg**
This PR fixes the `fp8` case, when `cutlass_mm` is not available. It contains the following fixes:
- Removes the padding for `fp8` `torch._scaled_mm`  in the `torch.compile` case, as branch specialization might not work correctly, and it makes fusion difficult.
- Implements redundant `slice` and `slice_scatter` elimination, which is implemented in PyTorch but does not cover all cases. It renames the `RedundantReshapesPass` to `NoopEliminationPass`.
- Minor custom pass improvements.

This PR is a pre-requisite PR to #10836, which enables `torch.compile` on AMD and uses the non-cutlass-fp8 path.

### CODE REVIEW

File: vllm/compilation/reshapes.py [hunk 1]
```diff
@@ -50,18 +66,48 @@ def __call__(self, graph: torch.fx.Graph):
...
+                input, dim_index, start, end = node.args[:4]
+                input_shape = input.meta["val"].shape
+                i_dim = input_shape[dim_index]
+
+                if start == 0 and self.dims_equivalent(end, i_dim):
+                    node.replace_all_uses_with(input)
+                    graph.erase_node(node)
+                    count += 1
+
+            elif is_func(node, torch.ops.aten.slice_scatter.default):
```
Comment (bnellnm) [inline vllm/compilation/reshapes.py:84]: Are these always the right ops to use?  e.g. is there a torch.ops.aten.slice.default or a torch.ops.aten.slice_scatter.Tensor?

Comment (ProExpertProg) [reply vllm/compilation/reshapes.py:84]: I haven't seen them so I am not sure - I just went off what I saw. The other overloads could be added easily if we ever see them in the graph

**Review by bnellnm — APPROVED** [review]

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py [hunk 2]
```diff
@@ -161,10 +162,14 @@ def apply_fp8_linear(
         # Note: we pad the input because torch._scaled_mm is more performant
         # for matrices with batch dimension > 16.
         # This could change in the future.
+        # We also don't pad when using torch.compile,
+        # as it breaks with dynamic shapes.
+        config = get_current_vllm_config().compilation_config
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/w8a8_utils.py:167]: Is this cached? It could be expensive each forward call

**Review by mgoin — APPROVED** [review]

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py (hunk 2 above)
Comment (ProExpertProg) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:167]: Yes, in eager mode this will get called on every forward pass, but it will only happen once when compiled. In eager mode there isn't really a better way that's still correct - the only way is to check the config context. I don't think this getter is significant but I haven't measured it.

File: tests/compile/test_fusion.py [hunk 3]
```diff
@@ -5,23 +5,25 @@
...
 
 from .backend import TestBackend
 
 
 class TestModel(torch.nn.Module):
 
-    def __init__(self, hidden_size: int, eps: float, static: bool, *args,
-                 **kwargs):
+    def __init__(self, hidden_size: int, eps: float, static: bool,
+                 cutlass_fp8: bool, *args, **kwargs):
```
Comment (tlrmchlsmth) [inline tests/compile/test_fusion.py:24]: please rename `cutlass_fp8` to something that indicates that it's a flag, e.g. `use_cutlass_fp8`

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py (hunk 2 above)
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:167]: We could passs in a `allow_input_padding` flag? and pass it in? I do think this is annoying though. I think it's woth it to do a quick check for performance regressions on a small model eager mode benchmark with `cutlass_scaled_mm` disabled?

File: vllm/compilation/reshapes.py
Comment (tlrmchlsmth) [inline vllm/compilation/reshapes.py:1]: rename this file `vllm/compilation/redundant_ops`? And similarly rename `RedundantReshapesPass` to `RedundantOpEliminationPass` or similar?

File: vllm/compilation/reshapes.py [hunk 4]
```diff
@@ -50,18 +66,48 @@ def __call__(self, graph: torch.fx.Graph):
...
+                    node.replace_all_uses_with(view)
+                    graph.erase_node(node)
+                    count += 1
+
         logger.debug("Removed %s no-op reshapes", count)
 
         self.dump_graph(graph, "after_reshapes")
         self.end_and_log()
 
+    def all_dims_equivalent(self, dims: Iterable[Union[int, torch.fx.Node]],
```
Comment (tlrmchlsmth) [inline vllm/compilation/reshapes.py:106]: Could you update the logs, maybe like the following?
```
        logger.debug("Removed %s no-op reshapes and slices", count)

        self.dump_graph(graph, "after_redundant_op_elimination")
```

**Review by tlrmchlsmth — COMMENTED** [review]
Looks good overall but I had a few minor comments

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py (hunk 2 above)
Comment (ProExpertProg) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py:167]: I think we'd have to pass that flag through the whole call stack though so I don't think it's worth it. I'll run a small model.

**Review by tlrmchlsmth — APPROVED** [review]
Thanks for the great work! LGTM assuming we don't see any performance regression

**Comment by ProExpertProg** [comment]
Yep will post perf numbers once I have them, thanks!

**Review by mgoin — APPROVED** [review]

**Comment by ProExpertProg** [comment]
> nit: noop elimination for slice errors when end = -1.
> 
> repro:
> 
> ```
> import torch
> 
> def dims_equivalent(dim, i_dim) -> bool:
>     # Case 1 and 2
>     if dim == i_dim or dim == -1:
>         return True
>     # Case 3
>     return isinstance(dim, torch.fx.Node) and dim.meta["val"] == i_dim
> 
> input = torch.randn((2,3,2))
> dim_index, start, end = 0, 0, -1
> 
> input_shape = input.shape
> 
> i_dim = input_shape[dim_index]
> 
> # following https://github.com/vllm-project/vllm/blob/main/vllm/compilation/noop_elimination.py#L79
> if start == 0 and dims_equivalent(end, i_dim):
>     is_noop = True
> else:
>     is_noop = False
> 
> if is_noop:
>     # input.shape: (2,3,2)
>     # torch.ops.aten.slice.Tensor(input, dim_index, start, end).shape: (1,3,2)
>     assert input.shape == torch.ops.aten.slice.Tensor(input, dim_index, start, end).shape
> ```
> 
> similarly for slice_scatter.

Great find - I didn't realize slice handles `end` differently. Could you file an issue for this please?

**Comment by BoyuanFeng** [comment]
issue: #17078

Btw, pytorch supports noop elimination for [view](https://github.com/pytorch/pytorch/pull/151095), [slice, and slice_scatter](https://github.com/pytorch/pytorch/pull/151175) now, which should be equivalent with `noop_elimination.py`.

**Comment by ProExpertProg** [comment]
Okay, sounds good, we can probably deprecate the pass, although it's nice to have an easy place to add noop transforms to unblock pattern matching in the short term before they fixes are upstreamed to torch.
